### DISCUSSION TITLE
How to create a rest-endpoint?

### DISCUSSION THREAD

**Discussion by PPetterssonSO** [SuperOffice/typescript-for-superoffice#13] [General]
Can i please have an example snippet on how to create a REST endpoint and fetch url parameters?

IE the JS version of the following code:
`%EJSCRIPT_START%
<%
String myValue = getCgiVariable("myValue");

Map data;
data.insert("You_have_sent: ", myValue);
print(data.toJson());
%>
%EJSCRIPT_END%`

So, when we do a GET to:
customer.fcgi?_sf=0&action=safeParse&includeId=SomeScript&key=SomeKey&myValue=JavaScriptRules

It would return
{"You_have_sent" : "JavaScriptRules"}

Regards
Pär

**Comment by ejfasting** [comment]
Hello @Nawksen 

There is an example on jumping in and out of ejscript-nuggets (%EJSCRIPT_START etc) here:
https://github.com/SuperOffice/javascript-for-superoffice/blob/main/Examples/HTML/HTMLPage-example.crmscript

Another example is this, which i haven't added to the repo (yet):
`%EJSCRIPT_START%
<%
context.result.body = JSON.stringify(context);
%>
%EJSCRIPT_END%`

This will pretty much prints out what's inside the context, which also contains the cgiVariables (as you are used to from 'old' crmscripts'), and should give you a better idea of what you have to work with and how you can pass inn variables to your scripts (as you describe). 

You work with json as you would in a 'normal' javascript.

I hope this answered your question :) If not, please do not hesitate to reply in this thread! Any question is appreciated, as it helps us share information to anyone who wants to get started and/or already have started testing out the functionality :)

//Eivind

**Comment by ehs5** [comment]
Cool! I tried it out for a bit, and found that headers are accessible through context.

Like in CRMScript, we can use REQUEST-METHOD to only allow  a certain request method like GET:

```
%EJSCRIPT_START%
<%
function print(text) {
    context.result.body += text;
}

function getHttpHeader(name) {
    return context.eventData.headers[name];
}

const method = getHttpHeader("REQUEST-METHOD");

if (method != "GET") {
    print("Invalid request method");
    
    // How to set status code other than 200?
    return;
}

// Request is GET, let's return something as JSON
const building = {
    name: "FA3",
    address: "Fyrstikkalléen 3A",
    zipcode: "0661",
    city: "Oslo"
}

print(JSON.stringify(building));

%>
%EJSCRIPT_END%
```

But how would we return a status code other than 200? And what about setting headers?


Best regards,
Espen Steen
Cloud Connection AS

**Reply by ejfasting** [reply]
Hi @ehs5 

I've created a backlog item to support this. Currently SuperOffice will always respond with `200 OK` , but i suppose what you want is to be able to do this:
```
%EJSCRIPT_START%
<%
context.result.status = 404; // <-
context.result.body = JSON.stringify(context);
%>
%EJSCRIPT_END%
```

You are able to set it, and its returned in the json (in this case), but the statusCode of the request does not reflect this. 

I will update this thread when I have news :) 

//E

**Reply by ehs5** [reply]
Yes, the end goal is to set status codes like 404 or whatever code we need. Exactly how that's achieved (setting a property on context, calling a method etc) is less important. But it is pretty important for us to be able to set status codes. Today we do this via the lib-http CRMScript library.

Thanks!

**Reply by PPetterssonSO** [reply]
Yeah that would be great, we are using lib-http in pretty much every solution we build (we do a lot of frontend -> Rest -> Backend solutions) that are dependent on the http lib :)
